src/intelligence_layer/core/model.py

```diff
@@ -408,21 +414,93 @@ def to_instruct_prompt(
+        rich_prompt = self.INSTRUCTION_PROMPT_TEMPLATE.to_rich_prompt(
             instruction=instruction, input=input, response_prefix=response_prefix
         )
 
+        prompt = rich_prompt.items[0]
+        ranges = rich_prompt.ranges
+
+        if not isinstance(prompt, Text):
+            raise ValueError("Text control only valid for text prompts.")
+
+        text_controls: list[TextControl] = []
+
+        if instruction_controls:
+            shifted_instruction_controls = self._shift_instruction_control_ranges(
+                instruction,
+                instruction_controls,
+                rich_prompt,
+            )
+            for shifted_input_control_range in shifted_instruction_controls:
+                text_controls.append(shifted_input_control_range)
+
+        if input_controls and input:
+            shifted_input_control_ranges = self._shift_input_control_ranges(
+                input, input_controls, rich_prompt
+            )
+            for shifted_input_control_range in shifted_input_control_ranges:
+                text_controls.append(shifted_input_control_range)
+
+        prompt_with_controls = Prompt.from_text(prompt.text, text_controls)
+        return RichPrompt(items=prompt_with_controls.items, ranges=ranges)
+
+    def _shift_instruction_control_ranges(
+        self,
+        instruction: str,
+        instruction_controls: Sequence[TextControl],
+        rich_prompt: RichPrompt,
+    ) -> Sequence[TextControl]:
+        text_controls = []
+        instruction_start = self._get_text_control_start_index(
+            rich_prompt, "instruction"
+        )
+        for control in instruction_controls:
+            if control.start + control.length > len(instruction):
+                raise ValueError(
+                    f"TextControl is out of bounds for instruction {instruction}"
+                )
+            text_controls.append(
+                replace(control, start=control.start + instruction_start)
+            )
+        return text_controls
+
+    def _shift_input_control_ranges(
+        self, input: str, input_controls: Sequence[TextControl], rich_prompt: RichPrompt
+    ) -> Sequence[TextControl]:
+        input_start = self._get_text_control_start_index(rich_prompt, "input")
+        text_controls = []
+        for control in input_controls:
+            if control.start + control.length > len(input):
+                raise ValueError(f"TextControl is out of bounds for input {input}")
+            text_controls.append(replace(control, start=control.start + input_start))
+        return text_controls
+
+    def _get_text_control_start_index(
+        self, rich_prompt: RichPrompt, control_type: str
+    ) -> int:
+        prompt_ranges = rich_prompt.ranges.get(control_type)
+        assert prompt_ranges is not None
+        assert (
+            len(prompt_ranges) == 1
+        )  # There should always only be one prompt range per control type.
```

Comment:
you could make this an assert with message: `assert False, "Oh no! This assertion failed!"`

Reply:
Done, good idea!